sql: detect old sentinel values correctly

vivekmenezes opened the pull request:
When a new column gets added to Family 0 it is set as the
default_column_id for family 0. When an old sentinel value is
read using this new family descriptor it should ignore
default_column_id. The default_column_id is
only be used for non 0 family values. Family 0 values are assumed
to be either sentinel values or of type VALUE_TUPLE.

Family 0 is always stored as a VALUE_TUPLE type and unfortunately
sentinel values are not stored as this type. When a sentinel value
is updated to include a non primary column, it is correctly upgraded
to a non-sentinel value of type VALUE_TUPLE.

Added TestParseSentinelValueWithNewColumnInSentinelFamily

fixes #12415

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12475)
<!-- Reviewable:end -->

vivekmenezes committed:
sql: detect old sentinel values correctly

When a new column gets added to Family 0 it is set as the
default_column_id for family 0. When an old sentinel value is
read using this new family descriptor it should ignore
default_column_id. The default_column_id is
only be used for non 0 family values. Family 0 values are assumed
to be either sentinel values or of type VALUE_TUPLE.

Family 0 is always stored as a VALUE_TUPLE type and unfortunately
sentinel values are not stored as this type. When a sentinel value
is updated to include a non primary column, it is correctly upgraded
to a non-sentinel value of type VALUE_TUPLE.

Added TestParseSentinelValueWithNewColumnInSentinelFamily

fixes #12415

tamird commented:
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/schema_changer_test.go, line 1412 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZNKbqNTuNqc3ceGNMD:-KZNKbqNTuNqc3ceGNME:bohjhlg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd061cb935f6961e2f11d71d7c66db9c64b2ed75/pkg/sql/schema_changer_test.go#L1412)):*
> ```Go
> 	}
> 	const eCount = maxValue + 1
> 	var val interface{}
> ```

why can't you make this a `*int`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/schema_changer_test.go, line 1412 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZNKbqNTuNqc3ceGNMD:-KZNxy7WtTVVk8Y4FifJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd061cb935f6961e2f11d71d7c66db9c64b2ed75/pkg/sql/schema_changer_test.go#L1412)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why can't you make this a `*int`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/schema_changer_test.go, line 1354 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZO55AGcs9Mk9ckxn-o:-KZO55AGcs9Mk9ckxn-p:b-7ymwdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/86c4c69e7de31df6f99217b5ff7da037ceee11f7/pkg/sql/schema_changer_test.go#L1354)):*
> ```Go
> 	const maxValue = 10
> 	inserts := make([]string, maxValue+1)
> 	for i := 0; i < maxValue+1; i++ {
> ```

`for i := range inserts`

---

*[pkg/sql/schema_changer_test.go, line 1361 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZO5MfuS6S902DlD98R:-KZO5MfuS6S902DlD98S:bfhaewu) ([raw file](https://github.com/cockroachdb/cockroach/blob/86c4c69e7de31df6f99217b5ff7da037ceee11f7/pkg/sql/schema_changer_test.go#L1361)):*
> ```Go
> 	}
> 
> 	// Downgrade table to a time when there were no column families.
> ```

but...why? this comment should explain the rationale

---

*[pkg/sql/schema_changer_test.go, line 1374 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZO5XSBVSLSi3U0rOob:-KZO5XSBVSLSi3U0rOoc:b-h3oupi) ([raw file](https://github.com/cockroachdb/cockroach/blob/86c4c69e7de31df6f99217b5ff7da037ceee11f7/pkg/sql/schema_changer_test.go#L1374)):*
> ```Go
> 	}
> 
> 	// Convert table data to sentinel values
> ```

ditto. perhaps this one and the above should be covered by the same comment

---

*[pkg/sql/schema_changer_test.go, line 1390 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZO5bt6iGA-eBUgocR9:-KZO5bt6iGA-eBUgocRA:b-1kl6fq) ([raw file](https://github.com/cockroachdb/cockroach/blob/86c4c69e7de31df6f99217b5ff7da037ceee11f7/pkg/sql/schema_changer_test.go#L1390)):*
> ```Go
> 	}
> 
> 	// Add column
> ```

not helpful

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/schema_changer_test.go, line 1354 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZO55AGcs9Mk9ckxn-o:-KZPF5l1TnMr7wNR1-Uq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/86c4c69e7de31df6f99217b5ff7da037ceee11f7/pkg/sql/schema_changer_test.go#L1354)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`for i := range inserts`
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 1361 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZO5MfuS6S902DlD98R:-KZPF_70X7JQ9FepS87E:b-1cleu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/86c4c69e7de31df6f99217b5ff7da037ceee11f7/pkg/sql/schema_changer_test.go#L1361)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

but...why? this comment should explain the rationale
</blockquote></details>

This was actually not needed, so I've taken it out.

---

*[pkg/sql/schema_changer_test.go, line 1374 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZO5XSBVSLSi3U0rOob:-KZPGH206wh1Faj-iKk0:bwo45yl) ([raw file](https://github.com/cockroachdb/cockroach/blob/86c4c69e7de31df6f99217b5ff7da037ceee11f7/pkg/sql/schema_changer_test.go#L1374)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto. perhaps this one and the above should be covered by the same comment
</blockquote></details>

I've changed this comment to:
```
// Convert table data to sentinel values. This is done to make
// the table appear like it were written in the past when
// cockroachdb used to write sentinel values for each table row.
```

---

*[pkg/sql/schema_changer_test.go, line 1390 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZO5bt6iGA-eBUgocR9:-KZPHLg3p9hSkan07Kyd:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/86c4c69e7de31df6f99217b5ff7da037ceee11f7/pkg/sql/schema_changer_test.go#L1390)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

not helpful
</blockquote></details>

fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/schema_changer_test.go, line 1345 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZRyIxLRpK-sZimtZ6c:-KZRyIxLRpK-sZimtZ6d:bmv1x9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcf1cb0c2cb43a32b8402f4cb9f5c1ff86d44d/pkg/sql/schema_changer_test.go#L1345)):*
> ```Go
> CREATE DATABASE t;
> CREATE TABLE t.test (
> 	k INT PRIMARY KEY
> ```

this test should be using the explicit family syntax everywhere since it cares very specifically about how columns get allocated to families

so this line should have `FAMILY f1 (k)` and the alter table below should have `FAMILY f1` on the end. https://www.cockroachlabs.com/docs/column-families.html

---

*[pkg/sql/schema_changer_test.go, line 1367 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZS4U6qdkrY9zfMRfAt:-KZS4U6qdkrY9zfMRfAu:b1sy3ii) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcf1cb0c2cb43a32b8402f4cb9f5c1ff86d44d/pkg/sql/schema_changer_test.go#L1367)):*
> ```Go
> 	// Convert table data to sentinel values. This is done to make
> 	// the table appear like it were written in the past when
> 	// cockroachdb used to write sentinel values for each table row.
> ```

can you mention that this overwrites the data put there by the insert? That threw me for a second

---

*[pkg/sql/schema_changer_test.go, line 1370 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZRyEa2JbzIxdx1BYfU:-KZRyEa2JbzIxdx1BYfV:blm5d8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcf1cb0c2cb43a32b8402f4cb9f5c1ff86d44d/pkg/sql/schema_changer_test.go#L1370)):*
> ```Go
> 	kvs, err := kvDB.Scan(
> 		context.TODO(),
> 		roachpb.Key(keys.MakeTablePrefix(uint32(tableDesc.ID))),
> ```

make the previous line a startKey and use startKey.PrefixEnd

---

*[pkg/sql/schema_changer_test.go, line 1393 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZS4IvDi1RLwsbRZWSF:-KZS4IvDi1RLwsbRZWSG:b-cv7trx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcf1cb0c2cb43a32b8402f4cb9f5c1ff86d44d/pkg/sql/schema_changer_test.go#L1393)):*
> ```Go
> 	}
> 
> 	// update one of the rows.
> ```

nit: Update

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-:-KZS4cUU8iUdfPeNVX61:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Is this manual crafting of "old style" data the right way to produce a test? Looking at this test, I can't see a reasonable way to validate that it's doing sensible and correct things.

---

Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-:-KZSolTNfs0PLaq6uWSq:bpa058h)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I've added a comment above the test describing what it tests. Thanks!

---

Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/schema_changer_test.go, line 1345 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZRyIxLRpK-sZimtZ6c:-KZU9ZI0wH-Vh1C1q2gq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcf1cb0c2cb43a32b8402f4cb9f5c1ff86d44d/pkg/sql/schema_changer_test.go#L1345)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary><blockquote>

this test should be using the explicit family syntax everywhere since it cares very specifically about how columns get allocated to families

so this line should have `FAMILY f1 (k)` and the alter table below should have `FAMILY f1` on the end. https://www.cockroachlabs.com/docs/column-families.html
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 1367 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZS4U6qdkrY9zfMRfAt:-KZU9ps6UQq6CjTJfQbx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcf1cb0c2cb43a32b8402f4cb9f5c1ff86d44d/pkg/sql/schema_changer_test.go#L1367)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary><blockquote>

can you mention that this overwrites the data put there by the insert? That threw me for a second
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 1370 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZRyEa2JbzIxdx1BYfU:-KZUA63rUdbdnJZmTE02:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcf1cb0c2cb43a32b8402f4cb9f5c1ff86d44d/pkg/sql/schema_changer_test.go#L1370)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary><blockquote>

make the previous line a startKey and use startKey.PrefixEnd
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 1393 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-KZS4IvDi1RLwsbRZWSF:-KZUA9oXxbFbslAfxJyv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9dbcf1cb0c2cb43a32b8402f4cb9f5c1ff86d44d/pkg/sql/schema_changer_test.go#L1393)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary><blockquote>

nit: Update
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12475#-:-KZUD852-uYzOm2fHlbZ:b732zhk)*
<!-- Sent from Reviewable.io -->
